### PR TITLE
fix: ToggleSwitch.Header is never visible

### PR DESCRIPTION
**Pull request by seanocali**
GitHub Issue (If applicable): #3437

## PR Type
- Bugfix
What kind of change does this PR introduce?

## What is the current behavior?
ToggleSwitch.Header.Visibility is always Collapsed and thus non-functional in non-UWP platforms

## What is the new behavior?
ToggleSwitch.Header functions across all platforms

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [X] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [X] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [X] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by jeromelaban** [comment]
Very nice fix ! It seems the Android test did not pass, I've restarted it to be validate if it's not hiding something else.

**Comment by seanocali** [comment]
> Very nice fix ! It seems the Android test did not pass, I've restarted it to be validate if it's not hiding something else.

Thanks, I'm trying to fix the test tonight.

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs [hunk 1]
```diff
@@ -187,13 +188,22 @@ public object Header
...
 		#endregion
 
 		protected override void OnApplyTemplate()
 		{
 			base.OnApplyTemplate();
 
+			_headerContentPresenter = GetTemplateChild("HeaderContentPresenter") as ContentPresenter;
 			_switchThumb = GetTemplateChild("SwitchThumb") as Thumb;
 			_switchKnob = GetTemplateChild("SwitchKnob") as FrameworkElement;
 			_switchKnobBounds = GetTemplateChild("SwitchKnobBounds") as FrameworkElement;
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs:209]:
```suggestion
			_switchKnobBounds = GetTemplateChild("SwitchKnobBounds") as FrameworkElement;
			
	        UpdateToggleState(false);
```

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs [hunk 2]
```diff
@@ -187,13 +188,22 @@ public object Header
...
-			DependencyProperty.Register("Header", typeof(object), typeof(ToggleSwitch), new PropertyMetadata(null)); 
+			DependencyProperty.Register("Header", typeof(object), typeof(ToggleSwitch), new PropertyMetadata(null, (s, e) => ((ToggleSwitch)s)?.OnHeaderChanged(e)));
+
+		private void OnHeaderChanged(DependencyPropertyChangedEventArgs e)
+		{
+			if (_headerContentPresenter != null)
+			{
+				_headerContentPresenter.Visibility = e.NewValue != null ? Visibility.Visible : Visibility.Collapsed;
+			}
+		}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs:199]: The change will work if the header is changed after the control template is set (after `OnApplyTemplate`), but if the Header is set before (e.g. set explicitly in XAML) you'll need to update the header from `OnApplyTemplate` as well:

```suggestion
		private void OnHeaderChanged(DependencyPropertyChangedEventArgs e)
		{
			if (_headerContentPresenter != null)
			{
				UpdateHeaderContentVisibility();
			}
		}

		private void UpdateHeaderContentVisibility() => _headerContentPresenter.Visibility = Header != null ? Visibility.Visible : Visibility.Collapsed;

```

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToggleSwitchTests/UnoSamples_Tests.ToggleSwitch.cs [hunk 3]
```diff
@@ -37,5 +37,28 @@ public void ToggleSwitch_TemplateReuseTest()
...
+		[AutoRetry]
+		public void ToggleSwitch_HeaderTest()
+		{
+			Run("UITests.Shared.Windows_UI_Xaml_Controls.ToggleSwitchControl.ToggleSwitch_Header");
+
+			var toggleSwitchWithHeader = _app.Marked("toggleSwitchWithHeader");
+			_app.WaitForElement(toggleSwitchWithHeader);
+
+			var toggleSwitchHeader = toggleSwitchWithHeader.Marked("toggleSwitchWithHeader");
+			_app.WaitForElement(toggleSwitchHeader);
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToggleSwitchTests/UnoSamples_Tests.ToggleSwitch.cs:51]: This is not needed, as `toggleSwitchWithHeader` cannot be named in the XAML (it's a property, not an actual visual tree element).
```suggestion
```

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToggleSwitchTests/UnoSamples_Tests.ToggleSwitch.cs [hunk 4]
```diff
@@ -37,5 +37,28 @@ public void ToggleSwitch_TemplateReuseTest()
...
+
+			var toggleSwitchHeaderContentTextBlock = _app.Marked("toggleSwitchHeaderContent");
+			_app.WaitForElement(toggleSwitchHeaderContentTextBlock);
+
+			var toggleSwitchNoHeader = _app.Marked("toggleSwitchNoHeader");
+			_app.WaitForElement(toggleSwitchNoHeader);
+
+			Assert.AreEqual("Visible", toggleSwitchWithHeader.GetDependencyPropertyValue("Visibility").ToString());
+			Assert.AreEqual("Test ToggleSwitch Header", toggleSwitchHeaderContentTextBlock.GetDependencyPropertyValue("Text").ToString());
+			Assert.IsNull(toggleSwitchNoHeader.GetDependencyPropertyValue("Header.Content"));
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/ToggleSwitchTests/UnoSamples_Tests.ToggleSwitch.cs:61]:
```suggestion
			Assert.AreEqual("", toggleSwitchNoHeader.GetDependencyPropertyValue("Header"));
```

When testing for this property, the UI Testing API cannot provide "null", but only a string representation of the property. Also, the `GetDependencyPropertyValue` only takes one property of the owner, and does not behave lie a binding path. Testing for `Header` here is enough.

File: src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs (hunk 2 above)
Comment (seanocali) [reply src/Uno.UI/UI/Xaml/Controls/ToggleSwitch.cs:199]: Yes I discovered this is as ApplyTemplate fires after OnHeaderChanged when running WebAssembly.

Initially I just recreated what already exists for Slider.Header, but clearly more needs to be done. Building and debugging in VS is still a challenge so it's taking awhile.

**Comment by seanocali** [comment]
Thanks for the feedback. I'll make it work.

**Comment by jeromelaban** [comment]
Make sure to add the shared project files, otherwise the tests will not pass.

**Comment by seanocali** [comment]
It looked like the empty header was actually returning null not an empty string and caused the test to fail.  I just removed the no-header test for now.  Is it even necessary?

**Comment by seanocali** [comment]
No idea why the latest didn't pass.

**Comment by seanocali** [comment]
Can someone check the snapshot that failed on Azure?

**Comment by jeromelaban** [comment]
That's fine it's an expected issue. Thanks again for your contribution!

**Pull request merged**